bot/cogs/moderation/incidents.py

```diff
@@ -0,0 +1,334 @@
+          relay the incident message to #incidents-archive
+        * If relay successful, delete original message
+        * See: `on_raw_reaction_add`
+
+    Please refer to function docstrings for implementation details.
+    """
+
+    def __init__(self, bot: Bot) -> None:
+        """Prepare `event_lock` and schedule `crawl_task` on start-up."""
+        self.bot = bot
+
+        self.event_lock = asyncio.Lock()
+        self.crawl_task = self.bot.loop.create_task(self.crawl_incidents())
+
+    async def crawl_incidents(self) -> None:
+        """
+        Crawl #incidents and add missing emoji where necessary.
+
+        This is to catch-up should an incident be reported while the bot wasn't listening.
+        After adding each reaction, we take a short break to avoid drowning in ratelimits.
+
+        Once this task is scheduled, listeners that change messages should await it.
+        The crawl assumes that the channel history doesn't change as we go over it.
+        """
+        await self.bot.wait_until_guild_available()
+        incidents: discord.TextChannel = self.bot.get_channel(Channels.incidents)
+
+        # Limit the query at 50 as in practice, there should never be this many messages,
+        # and if there are, something has likely gone very wrong
+        limit = 50
+
+        # Seconds to sleep after adding reactions to a message
+        sleep = 2
+
+        log.debug(f"Crawling messages in #incidents: {limit=}, {sleep=}")
+        async for message in incidents.history(limit=limit):
+
+            if not is_incident(message):
+                log.trace("Skipping message: not an incident")
+                continue
+
+            if has_signals(message):
+                log.trace("Skipping message: already has all signals")
+                continue
+
+            await add_signals(message)
+            await asyncio.sleep(sleep)
+
+        log.debug("Crawl task finished!")
+
+    async def archive(self, incident: discord.Message, outcome: Signal) -> bool:
+        """
+        Relay `incident` to the #incidents-archive channel.
+
+        The following pieces of information are relayed:
+            * Incident message content (clean, pingless)
+            * Incident author name (as webhook author)
+            * Incident author avatar (as webhook avatar)
+            * Resolution signal (`outcome`)
+
+        Return True if the relay finishes successfully. If anything goes wrong, meaning
+        not all information was relayed, return False. This signals that the original
+        message is not safe to be deleted, as we will lose some information.
+        """
+        log.debug(f"Archiving incident: {incident.id} with outcome: {outcome}")
+        try:
+            # First we try to grab the webhook
+            webhook: discord.Webhook = await self.bot.fetch_webhook(Webhooks.incidents_archive)
+
+            # Now relay the incident
+            message: discord.Message = await webhook.send(
```

Comment:
Not sure is this type hint required here.

Reply:
`Webhook.send` will return `None` if unless explicitly `wait=True`. [See here](https://discordpy.readthedocs.io/en/latest/api.html#discord.Webhook.send).

The annotation is there because we have `wait=True`, so we can make a promise of what the type will be. I like these hints because then PyCharm tells me what methods there are, and what params they take. The `d.py` lib isn't annotated, so this cannot be inferred automatically.

Reply:
OK, now I understand. Just as a note that dpy have PR opened about adding type hinting.